widgets/indicator/indicator.go

```diff
@@ -0,0 +1,206 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package Indicator is a widget that displays the status of a background operation
+// or shows whether a value is (on/off)
+package indicator
+
+import (
+	"errors"
+	"fmt"
+	"image"
+	"sync"
+	"time"
+
+	"github.com/mum4k/termdash/align"
+	"github.com/mum4k/termdash/internal/alignfor"
+	"github.com/mum4k/termdash/internal/area"
+	"github.com/mum4k/termdash/internal/canvas"
+	"github.com/mum4k/termdash/internal/canvas/braille"
+	"github.com/mum4k/termdash/internal/draw"
+	"github.com/mum4k/termdash/internal/runewidth"
+	"github.com/mum4k/termdash/terminal/terminalapi"
+	"github.com/mum4k/termdash/widgetapi"
+)
+
+// Indicator displays the progress of an operation by filling a partial circle and
+// The circle can has a "hole" in the middle, which is where the name comes from.
+// Implements widgetapi.Widget. This object is thread-safe.
+type Indicator struct {
+	// status is the current indication that will be drawn.
+	status bool
+	// mu protects the Indicator.
+	mu sync.Mutex
+
+	// opts are the provided options.
+	opts *options
+}
+
+// New returns a new Indicator.
+func New(opts ...Option) (*Indicator, error) {
+	opt := newOptions()
+	for _, o := range opts {
+		o.set(opt)
+	}
+	if err := opt.validate(); err != nil {
+		return nil, err
+	}
+	return &Indicator{
+		opts: opt,
+	}, nil
+}
+
+func (i *Indicator) Flash() error {
+	if i.status == true {
+		i.status = false
+	} else {
+		i.status = true
+	}
+	return nil
+}
+
+func (i *Indicator) Pulse() error {
+	for n := 0; n < 50; n++ {
+		if i.status == true {
+			i.status = false
+		} else {
+			i.status = true
+		}
+		time.Sleep(250 * time.Millisecond)
+	}
+	return nil
+}
+
+func (i *Indicator) drawLabel(cvs *canvas.Canvas, labelAr image.Rectangle) error {
+	start, err := alignfor.Text(labelAr, i.opts.label, i.opts.labelAlign, align.VerticalMiddle)
+	if err != nil {
+		return err
+	}
+	if err := draw.Text(
+		cvs, i.opts.label, start,
+		draw.TextOverrunMode(draw.OverrunModeThreeDot),
+		draw.TextMaxX(labelAr.Max.X),
+		draw.TextCellOpts(i.opts.labelCellOpts...),
+	); err != nil {
+		return err
+	}
+	return nil
+}
+
+// Draw draws the Indicator widget onto the canvas.
+// Implements widgetapi.Widget.Draw.
+func (i *Indicator) Draw(cvs *canvas.Canvas, meta *widgetapi.Meta) error {
+	i.mu.Lock()
+	defer i.mu.Unlock()
+
+	needAr, err := area.FromSize(i.minSize())
+	if err != nil {
+		return err
+	}
+	if !needAr.In(cvs.Area()) {
+		return draw.ResizeNeeded(cvs)
+	}
+	var indAr, labelAr image.Rectangle
+	if len(i.opts.label) > 0 {
+		i, l, err := indicatorAndLabel(cvs.Area())
+		if err != nil {
+			return err
+		}
+		indAr = i
+		labelAr = l
+
+	} else {
+		indAr = cvs.Area()
+	}
+	var t string
+	// split the indicator and label
+	// checked: \u25C9
+	// unchecked: \u25EF
+	// line: \u2500─────────────
+	if i.status == true {
+		t = "\u25C9"
```

Comment:
The next thing we should talk about is scaling. Most of the Termdash widgets scale in size as the container grows. If we just implement the drawing as a single character, it will always have a fixed size.

This is really up to you as the author of the widget. Do you think we should have this as a fixed size widget (the current implementation)? Or should we use the character only when the size of the container is small and use two concentric circles on a braille canvas as the size grows?

Reply:
You raise a really good point, I really like your idea of scaling up to be a braille canvas widget and scaling down to be a unicode character... I'll give that a shot and see if it looks ok.

Reply:
Sounds good, thanks @keithknott26.

To be noted - we can also agree to add the "scaling" at a later time, it isn't a feature that has to be in from day one. Completely up to you.

Reply:
When you have a moment , have a look at the indicator demo and let me know if you think that looks ok. If so I'll try to figure out how to display the scaled version and then the unicode version when things get below 3,3

Reply:
Also, I was hoping I to apply some mouse handling to this indicator as well since it now has the on/off/toggle methods. I'd like your suggestion on implementing a type of "onclick" event handler, it might be beyond my capabilities but I think it would be a good addition to this widget

Reply:
I have checked out the demo and it looks good! The scaling looks good.. We should probably add logic that chooses one or the other based on the canvas size, i.e. when it is below {3, 3} we should only use the character version, otherwise we should use the scaled version.

Secondly we might want to add an Option that sets the maximum indicator size as a diameter and when 1-3 is passed, it will only use the character version. So that the user can limit the scaling. What do you think?

When it comes to handling mouse events - that actually isn't too hard. There is an implementation of a mouse state machine (tracks clicks and determines when a full click happened). A full click is when the button is both pressed and released in the widget's area. You can look at the button widget and how the FSM is used. Before we go there - what do you envision this used as? What should happen when the user clicks on the indicator?

Reply:
Very good suggestions, I've included these changes in the latest commit.